### DISCUSSION TITLE
How to connect a flask docker container to planetscale?

### DISCUSSION THREAD

**Discussion by abstractprototype** [planetscale/discussion#455] [Documentation] [answered]
I am currently using a Dockerfile to build the image, however I get an error when running the container. I am using windows and downloaded a cacert.pem file to be used for SSL. Am I missing something? There seem to be solutions for prisma and node js but none for Flask after searching.
![image](https://user-images.githubusercontent.com/61066073/235614631-c6a9ca65-2a2e-4acc-ad27-92ba119caa45.png)
![image](https://user-images.githubusercontent.com/61066073/235631080-d142ca14-1c15-47e1-930a-562ce1390544.png)
![image](https://user-images.githubusercontent.com/61066073/235614797-4395d239-510b-4f12-9bba-ecb05f890e97.png)
![image](https://user-images.githubusercontent.com/61066073/235614891-e88cffdd-e52b-4de1-982b-08679d417b5a.png)

**Comment by jgreet** [comment, marked as answer]
Hello,

In the Docker image you're using, you shouldn't need to download a CA bundle, you can point to the system CA bundle at this location: `/etc/ssl/certs/ca-certificates.crt`.

You also don't want to specify `ssl_cert` or `ssl_key`, only the system CA bundle should be necessary to verify the PlanetScale server.

Beyond that, you also want to be sure you're using `mysql-connector-python` and not the outdated `mysql-connector` package.

Here's a simplified version of our `mysql-connector-python` example that I was able to run in the same Docker image you're starting with:

From https://planetscale.com/blog/connect-to-a-mysql-database-in-python#connect-to-mysql-with-the-mysql-connector

```python
from mysql.connector import Error
import mysql.connector

connection = mysql.connector.connect(
host="aws.connect.psdb.cloud",
database="your-database-name",
user="redacted",
password="redacted",
ssl_ca="/etc/ssl/certs/ca-certificates.crt"
)

try:
    if connection.is_connected():
        cursor = connection.cursor()
    cursor.execute("select @@version ")
    version = cursor.fetchone()
    if version:
        print('Running version: ', version)
    else:
        print('Not connected.')
except Error as e:
    print("Error while connecting to MySQL", e)
finally:
    connection.close()
```

```shell
root@08925b4d8f6e:~# python test.py
Running version:  ('8.0.21-Vitess',)
```

**Reply by jgreet** [reply]
I should add I'm just guessing you're using `mysql-connector-python` from the code in your screenshot. If you're using a different driver the same advice applies, use the system CA bundle which you should only need to specify once.

Also, I'm not sure what you're trying to do here but it looks like you're starting the Flask app with our domain on the standard MySQL port. That's probably not what you actually want.

![](https://user-images.githubusercontent.com/61066073/235614797-4395d239-510b-4f12-9bba-ecb05f890e97.png)

**Reply by abstractprototype** [reply]
Oh ok I see that I don't need to download any CA files, I just need to add that one line of code for ssl_ca? I read on the documentation that windows needs to download a cacert.pem to be used.
I am using the latest mysql-connector-python. I also change the way I start my Flask app back to host='0.0.0.0' and port=5001. On the Dockerfile I also removed the 'ENV SSL_CERT_FILE'. 

However, I am still getting the same error. I will attach my updated code with your suggestions and the mysql-connector-python example.
The commands I use to build my image and run the container are ```docker build -t flask-server .``` and ```docker run -p 5001:5001 flask-server```. Could this error possibly be something I need to enable on my planetscale database account?
![image](https://user-images.githubusercontent.com/61066073/236134993-4492cab8-38f2-403f-8ae2-dd9fffbf5c61.png)
![image](https://user-images.githubusercontent.com/61066073/236134675-d0e9c139-f3cb-4781-864f-271293f3b6ec.png)
![image](https://user-images.githubusercontent.com/61066073/236135480-b5f305ec-6a2a-48bd-b10f-2f6f2777ddf9.png)

**Reply by jgreet** [reply]
![image](https://user-images.githubusercontent.com/391331/236539807-8ccd1c18-9669-429a-98c8-632245cd5e08.png)

This error indicates that your driver is limited to the outdated, insecure TLSv1. Perhaps you can specify a `tls-versions` argument setting it to 1.3 but you shouldn't have to, recent releases will use a current TLS version by default.

Upgrading the version of `mysql-connector-python` the app is using (and double checking again that you're nothing using `mysql-connector`, basically an unfortunate fork which is outdated and limited to TLSv1) is probably the answer.

**Reply by abstractprototype** [reply]
Thank you it seems I forgot to include mysql-connector-python on my requirements.txt file. I can run my container now.

**Reply by jgreet** [reply]
Awesome! Glad to hear it 🎉

**Comment by TheCliffJumper** [comment]
I am using windows os and getting :-
FileNotFoundError: [Errno 2] No such file or directory
: this error what should i do please help

from flask import Flask, render_template, request, redirect, url_for
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

from datetime import datetime


app = Flask(__name__)

app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql+pymysql://vgzq84iy5bw4tvwqessi:pscale_pw_****@aws.connect.psdb.cloud/todo?ssl_ca=/etc/ssl/certs/ca-certificates.crt'
db = SQLAlchemy(app)

class Todo(db.Model):
    __tablename__ = 'todolist'

    sno = db.Column(db.Integer, primary_key=True)
    title = db.Column(db.String(250), nullable=False)
    descr = db.Column(db.String(500), nullable=False)
    date_time = db.Column(db.DateTime, default=datetime.utcnow)

    def __init__(self, title, descr):
        self.title = title
        self.descr = descr


@app.route('/', methods=['GET', 'POST'])
def home():
    if request.method == 'POST':
        title = request.form.get('title')
        descr = request.form.get('descr')

        engine = create_engine('mysql+pymysql://vgzq84iy5bw4tvwqessi:pscale_pw_****@aws.connect.psdb.cloud/todo?ssl_ca=/etc/ssl/certs/ca-certificates.crt')
        Session = sessionmaker(bind=engine)
        session = Session()

        new_record = Todo(title=title, descr=descr)
        session.add(new_record)
        session.commit()

        session.close()

        return redirect(url_for('home'))

    return render_template('index.html')


@app.route('/update')
def update():
    return "update"


@app.route('/delete')
def delete(sno):
    return "all products"


if __name__ == "__main__":
    app.run(host='0.0.0.0', debug=True)

**Reply by jgreet** [reply]
I edited your posts to remove the passwords. Be sure to revoke that password if the automated credential scanner hasn't caught it already.

https://planetscale.com/docs/concepts/connection-strings#deleting-a-password

Since you're using Windows, this is likely the problem `ssl_ca=/etc/ssl/certs/ca-certificates.crt`.  That's the correct file path for several Linux distributions but probably invalid on your computer.

On Windows, if you need to set `ssl_ca` at all, you'll want to set it to the absolute path of the certificate bundle you downloaded:

https://planetscale.com/docs/concepts/secure-connections#windows

**Reply by TheCliffJumper** [reply]
Grateful for your assistance. Thank you for your invaluable support!
